src/Service.Tests/Unittests/MultiSourceQueryExecutionUnitTests.cs

```diff
@@ -146,15 +121,110 @@ public async Task TestMultiSourceQuery()
+            sqlMetadataProviderDb1.Setup(x => x.GetLinkingEntities()).Returns(new Dictionary<string, Entity>());
+            sqlMetadataProviderDb2.Setup(x => x.GetLinkingEntities()).Returns(new Dictionary<string, Entity>());
+            sqlMetadataProviderDb1.Setup(x => x.GetDatabaseType()).Returns(DatabaseType.MySQL);
+            sqlMetadataProviderDb2.Setup(x => x.GetDatabaseType()).Returns(DatabaseType.CosmosDB_NoSQL);
+            metadataProviderFactory.Setup(x => x.GetMetadataProvider(DATA_SOURCE_NAME_1)).Returns(sqlMetadataProviderDb1.Object);
+            metadataProviderFactory.Setup(x => x.GetMetadataProvider(DATA_SOURCE_NAME_2)).Returns(sqlMetadataProviderDb2.Object);
+            authorizationService.Setup(x => x.AuthorizeAsync(It.IsAny<ClaimsPrincipal>(), It.IsAny<object>(), It.IsAny<IEnumerable<IAuthorizationRequirement>>())).Returns(Task.FromResult(AuthorizationResult.Success()));
+            RequestValidator requestValidator = new(metadataProviderFactory.Object, provider);
+
+            // Setup REST Service
+            RestService restService = new(
+                queryEngineFactory.Object,
+                mutationEngineFactory.Object,
+                metadataProviderFactory.Object,
+                httpContextAccessor.Object,
+                authorizationService.Object,
+                provider,
+                requestValidator);
+
+            // client is mapped as belonging to the sql data source.
+            // customer is mapped as belonging to the cosmos data source.
+            await restService.ExecuteAsync(ENTITY_NAME_1, EntityActionOperation.Read, null);
+
+            Assert.AreEqual(1, sqlQueryEngine.Invocations.Count, "Sql query engine should be invoked for multi-source query as entity belongs to sql db.");
+            Assert.AreEqual(0, cosmosQueryEngine.Invocations.Count, "Cosmos query engine should not be invoked for multi-source query as entity belongs to sql db.");
+            sqlQueryEngine.Verify(x => x.ExecuteAsync(It.Is<FindRequestContext>(ctx => ctx.EntityName == ENTITY_NAME_1)), Times.Once);
+            cosmosQueryEngine.Verify(x => x.ExecuteAsync(It.IsAny<FindRequestContext>()), Times.Never);
+            IActionResult result = await restService.ExecuteAsync(ENTITY_NAME_2, EntityActionOperation.Read, null);
+            Assert.AreEqual(1, cosmosQueryEngine.Invocations.Count, "Cosmos query engine should be invoked for multi-source query as entity belongs to cosmos db.");
+            Assert.AreEqual(1, sqlQueryEngine.Invocations.Count, "Sql query engine should not be invoked again for multi-source query as entity2 belongs to cosmos db.");
+            sqlQueryEngine.Verify(x => x.ExecuteAsync(It.Is<FindRequestContext>(ctx => ctx.EntityName == ENTITY_NAME_1)), Times.Once);
+            cosmosQueryEngine.Verify(x => x.ExecuteAsync(It.Is<FindRequestContext>(ctx => ctx.EntityName == ENTITY_NAME_2)), Times.Once);
```

Comment:
nit add newline between Asserts from first "Act" operation. It's not easy to read that there are, in fact, two calls to the system under test (SUT) here. 
1. First test -> entity1
2. Second test -> entity2

In the first test, you check CosmosQueryEngine is never invoked, then you verify if CosmosQueryEngine.ExecuteAsync is never invoked. Isn't the second check redundant since zero invocations of CosmosQueryEngine implies zero invocations of CosmosQueryEngine.ExecuteAsync?

Same situation for the second test but you're redundantly checking that SqlQueryEngine.ExecuteAsync is not invoked again after already ensuring SqlQueryENgine isn't invoked again.

Reply:
Makes sense. fixed

Reply:
Does my understanding of your validations reflect what you intended to test? If I'm incorrect, that's fine, just go ahead and add extra comments if needed.